storage: add test to prevent regression of #7600

tbg opened the pull request:
This took quite a bit of fiddling, but that version fails both test
and testrace almost instantly (<30iters) when the code added in #7672
is disabled (and even faster with more aggressive Raft tick intervals).
It does not catch the clobbering that could happen if
storeReplicaRaftReadyConcurrency were increased, at least not within 500
iterations.

Closes #7600.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7700)

<!-- Reviewable:end -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 5 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[storage/client_split_test.go, line 1091 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7700#-KMBJym2g3t0AvQpOwLz:-KMBJym2g3t0AvQpOwM-:416318335) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4d34ba9b824e3ae19d1e79729e461098910344/storage/client_split_test.go#L1091)):*
> ```Go
> 			// side in the split trigger was racing with the uninitialized
> 			// version for the same group, resulting in clobbered HardState).
> 			for term := uint64(1); term < storage.RaftInitialLogTerm; term++ {
> ```

Why `< RaftInitialLogTerm`? Isn't the same effect seen no matter what the term? (so we should set the number of iterations based on what we need to reproduce the error instead of tying it to this constant)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7700#-:-KMBKPBxvkNN79t6NvuO:1417088721)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: add test to prevent regression of #7600

This took quite a bit of fiddling, but that version fails both test
and testrace almost instantly (<30iters) when the code added in #7672
is disabled (and even faster with more aggressive Raft tick intervals).
It does not catch the clobbering that could happen if
storeReplicaRaftReadyConcurrency were increased, at least not within 500
iterations.

tschottdorf commented:
I just want to get in some iterations to increase the likelihood of the
race. I don't think there's any downside to going past the initial log
term, but the group will start at 6 and I don't want to unnecessarily
disturb it. Besides, whatever number I pick here will probably work about
the same. But I see your point that us changing the initial log term to 0
(which we could) shouldn't mess with this test. Changing to `5`.

-- 

-- Tobias

petermattis commented:
@tschottdorf I haven't tried to grok this test, but if you're adding iterations in the hopes of exercising a particular race I think a different approach would be better. Google had a small test utility library called `TestPoints` which were named hooks placed inside production code that test code could control. The production code (i.e. `Replica.splitTrigger`) would call `TestPoint.Reached()`. By default, this method is a no-op, but tests could configure individual test points to block, to get notified when the `TestPoint` was reached and to allow the code to continue processing. This allowed fine-grained control of unusual timings.

tschottdorf commented:
Mind filing an issue? I'm not interested in tackling this at this point.

On Fri, Jul 8, 2016 at 8:05 PM Peter Mattis <notifications@github.com>
wrote:

> @tschottdorf <https://github.com/tschottdorf> I haven't tried to grok
> this test, but if you're adding iterations in the hopes of exercising a
> particular race I think a different approach would be better. Google had a
> small test utility library called TestPoints which were named hooks
> placed inside production code that test code could control. The production
> code (i.e. Replica.splitTrigger) would call TestPoint.Reached(). By
> default, this method is a no-op, but tests could configure individual test
> points to block, to get notified when the TestPoint was reached and to
> allow the code to continue processing. This allowed fine-grained control of
> unusual timings.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/7700#issuecomment-231500275>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AE135Fbv4CssNT2XdwaOqfjDNCscM0ysks5qTuXYgaJpZM4JIMM->
> .
>
-- 

-- Tobias
